src/OSPSuite.Core/Domain/Builder/SimulationBuilder.cs

```diff
@@ -203,38 +266,131 @@ private void addMolecules(IMoleculeDependentBuilder builder, MoleculeList molecu
+            .Select(x => (propAccess(x.Module), x.MergeBehavior))
+            .ToList();
+
       public virtual IReadOnlyCollection<SimulationEntitySource> EntitySources => _entitySources;
 
-      internal MoleculeList MoleculeListFor(IMoleculeDependentBuilder builder) => _moleculeListCache[builder];
+      internal MoleculeList MoleculeListFor(IMoleculeDependentBuilder builder) => builder.MoleculeList;
 
       internal MoleculeBuilder MoleculeByName(string name) => _molecules[name];
 
-      internal class BuilderSource
+      internal class BuilderSource : BuilderSource<IEntity>
       {
-         public IEntity Builder { get; }
+         public BuilderSource(IEntity builder, IBuildingBlock buildingBlock) : base(builder, buildingBlock)
+         {
+         }
+      }
+
+      internal BuilderSource BuilderSourceFor(IEntity sourceEntity) => _builderSources[sourceEntity.Id];
+
+      internal class BuilderSource<T> where T : IEntity
+      {
+         public T Builder { get; }
          public IBuildingBlock BuildingBlock { get; }
 
-         public BuilderSource(IEntity builder, IBuildingBlock buildingBlock)
+         public BuilderSource(T builder, IBuildingBlock buildingBlock)
          {
             Builder = builder;
             BuildingBlock = buildingBlock;
          }
 
-         public override string ToString()
+         public void Deconstruct(out T builder, out IBuildingBlock buildingBlock)
          {
-            return $"{Builder.EntityPath()} - {BuildingBlock.Name}";
+            builder = Builder;
+            buildingBlock = BuildingBlock;
          }
       }
 
-      internal BuilderSource BuilderSourceFor(IEntity sourceEntity) => _builderSources[sourceEntity.Id];
+      internal class BuilderMergeInfo<T> where T : IBuilder
+      {
+         public BuilderSource<T> BaseBuilder { get; }
+         public IReadOnlyList<BuilderSource<T>> BuildersToMerge { get; }
+
+         /// <summary>
+         ///    Indicates that the base builder needs to be cloned before merging
+         /// </summary>
+         public bool RequiresClone => BuildersToMerge.Count > 0;
+
+         /// <summary>
+         ///    It is required to also clone each builder being merged if we have 2 or more builders to merge.
+         ///    This prevents cross-contamination between builders during sequential merges.
+         ///    Example: When merging B1, then B2 into finalBuilder:
+         ///    - Merge B1: B1's children/references are transferred to finalBuilder
+         ///    - Merge B2: B2's merge operations might modify children that came from B1
+         ///    - If B1 wasn't cloned, these modifications propagate back to the original building block through shared references
+         ///    With only 1 builder to merge, there are no subsequent merges to cause side effects, so cloning is not needed.
+         /// </summary>
+         public bool RequiresMergeClone => BuildersToMerge.Count >= 2;
```

Comment:
Hopefully this comments makes it clear

Reply:
I made some comments about a general sense I have when reading about merge/extend. Is there anywhere that the word **Extend** could replace **Merge**